adapters/smartyads/smartyads.go

```diff
@@ -0,0 +1,209 @@
+)
+
+// Implements Bidder interface.
+type SmartyAdsAdapter struct {
+	endpoint template.Template
+}
+
+func NewSmartyAdsBidder(endpointTemplate string) *SmartyAdsAdapter {
+	template, err := template.New("endpointTemplate").Parse(endpointTemplate)
+	if err != nil {
+		return nil
+	}
+	return &SmartyAdsAdapter{endpoint: *template}
+}
+
+func (a *SmartyAdsAdapter) CheckHasImps(request *openrtb.BidRequest) error {
+	if len(request.Imp) == 0 {
+		err := &errortypes.BadInput{
+			Message: "Missing Imp Object",
+		}
+		return err
+	}
+	return nil
+}
+
+func GetHeaders(request *openrtb.BidRequest) *http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+
+		if len(request.Device.Language) > 0 {
+			headers.Add("Accept-Language", request.Device.Language)
+		}
+
+		if request.Device.DNT != nil {
+			headers.Add("Dnt", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+
+	return &headers
+}
+
+func (a *SmartyAdsAdapter) MakeRequests(
+	openRTBRequest *openrtb.BidRequest,
+	reqInfo *adapters.ExtraRequestInfo,
+) (
+	requestsToBidder []*adapters.RequestData,
+	errs []error,
+) {
+
+	request := *openRTBRequest
+
+	if noImps := a.CheckHasImps(&request); noImps != nil {
+		return nil, []error{noImps}
+	}
+
+	var errors []error
+	var smartyadsExt *openrtb_ext.ExtSmartyAds
+	var err error
+
+	for i, imp := range request.Imp {
+		smartyadsExt, err = a.getImpressionExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			break
+		}
+		request.Imp[i].Ext = nil
+	}
+
+	if len(errors) > 0 {
+		return nil, errors
+	}
+
+	url, err := a.buildEndpointURL(smartyadsExt)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	return []*adapters.RequestData{{
+		Method:  http.MethodPost,
+		Body:    reqJSON,
+		Uri:     url,
+		Headers: *GetHeaders(&request),
+	}}, nil
+}
+
+func (a *SmartyAdsAdapter) getImpressionExt(imp *openrtb.Imp) (*openrtb_ext.ExtSmartyAds, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "ext.bidder not provided",
+		}
+	}
+	var smartyadsExt openrtb_ext.ExtSmartyAds
+	if err := json.Unmarshal(bidderExt.Bidder, &smartyadsExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "ext.bidder not provided",
+		}
+	}
+	return &smartyadsExt, nil
+}
+
+func (a *SmartyAdsAdapter) buildEndpointURL(params *openrtb_ext.ExtSmartyAds) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{Host: params.Host, SourceId: params.SourceId, AccountID: params.AccountID}
```

Comment:
Adapters may not allow the publisher to specify the entire endpoint domain. I can offer more specific suggestions if you could please give us more context on why a publisher would specify the host.

Reply:
Hello! @SyntaxNode , thanks for you review! Our company has different ns for different regions such as (US_EAST, EU, SGP) . We need to use for example endpoints ns1.example.com for EU and ns2.example.com for SGP. That is main reason when I use host variable.

Reply:
Should the region not be controlled by the PBS host (as in adapter config) rather than publisher, as the adapter host would control which endpoint is "closest"? Or if the region specific requirement is driven by which advertisers are active on an endpoint (east coat advertisers just want to bid on requests coming to the US_EAST node), then from your example have host = "ns1" and hardcode ".example.com" into your adapter's URL.

The reasoning behind this is we don't want PBS to be used as an open relay. For example, with an openended host setting, someone could have "host": "www.disney.com" and then flood PBS installs with requests which would then be forwarded to www.disney.com to try to overload their website. Given that Disney will see all the requests coming from PBS hosts rather than the original attacker, it will be most difficult or impossible to find the one responsible.

Reply:
@hhhjort , thanks for your answer ! I will change host to hardcode ".example.com" domain with ns1 as parameter! Thank you for clarification!